qiskit/aqua/operators/gradients/circuit_qfis/overlap_diag.py

```diff
@@ -0,0 +1,258 @@
+                         ) -> OperatorBase:
+        """
+        Args:
+            operator: The operator corresponding to the quantum state |ψ(ω)〉for which we compute
+                the QFI
+            params: The parameters we are computing the QFI wrt: ω
+
+        Returns:
+            ListOp where the operator at position k corresponds to QFI_k,k
+
+        Raises:
+            NotImplementedError: If a circuit is found such that one parameter controls multiple
+                gates, or one gate contains multiple parameters.
+
+        """
+
+        if not isinstance(operator, (CircuitOp, CircuitStateFn)):
+            raise NotImplementedError('operator must be a CircuitOp or CircuitStateFn')
+
+        circuit = operator.primitive
+
+        # Partition the circuit into layers, and build the circuits to prepare $\psi_i$
+        layers = self._partition_circuit(circuit)
+        if layers[-1].num_parameters == 0:
+            layers.pop(-1)
+
+        psis = [CircuitOp(layer) for layer in layers]
+        for i, psi in enumerate(psis):
+            if i == 0:
+                continue
+            psis[i] = psi @ psis[i - 1]
+
+        # TODO: make this work for other types of rotations
+        # NOTE: This assumes that each parameter only affects one rotation.
+        # we need to think more about what happens if multiple rotations
+        # are controlled with a single parameter.
+        generators = self._get_generators(params, circuit)
+
+        diag = []
+        for param in params:
+            if len(circuit._parameter_table[param]) > 1:
+                raise NotImplementedError("OverlapDiag does not yet support multiple "
+                                          "gates parameterized by a single parameter. For such "
+                                          "circuits use LinCombFull")
+
+            gate = circuit._parameter_table[param][0][0]
+
+            assert len(gate.params) == 1, "OverlapDiag cannot yet support gates with more than " \
+                                          "one parameter."
+
+            param_value = gate.params[0]
+            generator = generators[param]
+            meas_op = ~StateFn(generator)
+
+            # get appropriate psi_i
+            psi = [(psi) for psi in psis if param in psi.primitive.parameters][0]
+
+            op = meas_op @ psi @ Zero
+            if isinstance(param_value, ParameterExpression) and not isinstance(param_value,
+                                                                               Parameter):
+                expr_grad = DerivativeBase.parameter_expression_grad(param_value, param)
+                op *= expr_grad
+            rotated_op = PauliExpectation().convert(op)
+            diag.append(rotated_op)
+
+        grad_op = ListOp(diag, combo_fn=lambda x: np.diag(np.real([1 - y ** 2 for y in x])))
+        return grad_op
+
+    @staticmethod
+    def _partition_circuit(circuit):
```

Comment:
Is this the exact same method as in `overlap_block_diag`? If yes it should just use the method from there

Reply:
From what I remember the diagonal implementation really just computes the diagonal and the block-diagonal computes the Q_ij terms where ij represent all parameters in one layer. Thus, you have more then solely the diagonal QFI elements.

Reply:
the code looked the same and still works if they use the same code -- so I removed one of them